File: src/interpreter.cpp
#include <map>

#include "parser.h"
#include "interpreter.h"

// Common errors
static void noVarErr(std::string name) {
    throw std::runtime_error("No varible named " + name + " in current scope");
}

static void noFuncErr(std::string name) {
    throw std::runtime_error("No function named " + name + " in current scope");
}

static void badArgsErr(int got, int need) {
    throw std::runtime_error(
        "Wrong number of arguments, got " + std::to_string(got)
        + ", need " + std::to_string(need)
    );
}

std::shared_ptr<Val> Runner::eval(std::shared_ptr<ExprAST> &node) {
    // Literals
    if (CAST_TO(node.get(), StringAST*)) {
        std::string str = CAST_TO(node.get(), StringAST*)->str;
        return std::make_shared<Val>(str);
    } else if (CAST_TO(node.get(), NumberAST*)) {
        int num = CAST_TO(node.get(), NumberAST*)->num;
        return std::make_shared<Val>(num);
    } else if (CAST_TO(node.get(), DecimalAST*)) {
        float num = CAST_TO(node.get(), DecimalAST*)->num;
        return std::make_shared<Val>(num);
    } else if (CAST_TO(node.get(), BoolAST*)) {
        bool val = CAST_TO(node.get(), BoolAST*)->val;
        return std::make_shared<Val>(val);
    } else if (CAST_TO(node.get(), NoneAST*)) {
        return std::make_shared<Val>();
    }
    // Vars
    if (CAST_TO(node.get(), VarAST*)) {
        std::string name = CAST_TO(node.get(), VarAST*)->name;
        std::shared_ptr<Val> var = state.get_var(name);
        if (var)
            return var;
        // Error if no var
        noVarErr(name);
    }
    // Unary
    if (CAST_TO(node.get(), UnaryAST*)) {
        auto unary = std::reinterpret_pointer_cast<UnaryAST>(node);
        // Recursively call eval until it all works out
        std::shared_ptr<Val> val = eval(unary->expr);
        if (!val)
            return NULL;
        if (unary->op == "-") {
            *val = Val(0) - *val;
        } else if (unary->op == "!") {
            *val = !*val;
        } else if (unary->op == "++") {
            *val += 1;
        } else if (unary->op == "--") {
            *val -= 1;
        }
        return val;
    }
    // Binop
    if (CAST_TO(node.get(), BinopAST*)) {
        auto binop = std::reinterpret_pointer_cast<BinopAST>(node);
        // Recursively call eval until it all works out
        Val val = Val();
        std::string op = binop->op;
        if (
            op == "=" || op == "+=" || op == "-=" ||
            op == "*=" || op == "/="
        ) {
            std::shared_ptr<Val> right = eval(binop->right);
            // Left is a var, get the name
            std::string name = CAST_TO(binop->left.get(), VarAST*)->name;
            // Get var
            std::shared_ptr<Val> var = state.get_var(name);
            if (!var) {
                // No var, error.
                noVarErr(name);
            }
            // Do something
            if (op == "=")
                *var = *right;
            if (op == "+=")
                *var += *right;
            if (op == "-=")
                *var -= *right;
            if (op == "*=")
                *var *= *right;
            if (op == "/=")
                *var /= *right;
            return NULL;
        }
        std::shared_ptr<Val> left = eval(binop->left);
        // Binary ops, bit tricky due to shortchir
        if (op == "&&")
            val = !!*left && !!*eval(binop->right);
        else if (op == "||")
            val = !!*left || !!*eval(binop->right);
        else if (op == "^^")
            val = !!*left != !!*eval(binop->right);
        if (val != Val())
            return std::make_shared<Val>(val);
        // Normal ops
        std::shared_ptr<Val> right = eval(binop->right);
        if (op == "+")
            val = *left + *right;
        else if (op == "-")
            val = *left - *right;
        else if (op == "*")
            val = *left * *right;
        else if (op == "/")
            val = *left / *right;
        else if (op == "%")
            val = *left % *right;
        else if (op == "==")
            val = *left == *right;
        else if (op == "!=")
            val = *left != *right;
        else if (op == "==")
            val = *left == *right;
        else if (op == ">")
            val = *left > *right;
        else if (op == "<")
            val = *left < *right;
        else if (op == ">=")
            val = *left >= *right;
        else if (op == "<=")
            val = *left <= *right;
        return std::make_shared<Val>(val);
    }
    // Call
    if (CAST_TO(node.get(), CallAST*)) {
        auto call = std::reinterpret_pointer_cast<CallAST>(node);
        Val val = call_func(call);
        return std::make_shared<Val>(val);
    }
    // No value
    return NULL;
}

std::shared_ptr<Val> Runner::exec_single(std::shared_ptr<StmtAST> &node) {
    if (CAST_TO(node.get(), ExprAST*)) {
        // Eval expression
        auto expr = std::reinterpret_pointer_cast<ExprAST>(node);
        std::shared_ptr<Val> val = eval(expr);
        if (val && *val != Val() && isRepl && !inFunc)
            puts(val->toStr().c_str());
    } else if (CAST_TO(node.get(), BodyAST*)) {
        // Execute body
        return exec_body(node);
    } else if (CAST_TO(node.get(), FuncAST*)) {
        // Add function
        auto func = std::reinterpret_pointer_cast<FuncAST>(node);
        state.funcs[func->name] = func;
    } else if (CAST_TO(node.get(), LetAST*)) {
        // Add var
        auto var = std::reinterpret_pointer_cast<LetAST>(node);
        if (state.vars.find(var->name) != state.vars.end()) {
            // Var already exists in current scope, err
            throw std::runtime_error("Redefinition of " + var->name);
        }
        state.vars[var->name] = eval(var->val);
    } else if (CAST_TO(node.get(), IfAST*)) {
        // Execute if
        auto ifExpr = std::reinterpret_pointer_cast<IfAST>(node);
        std::shared_ptr<Val> cond = eval(ifExpr->cond);
        if (cond && !!*cond) {
            // It's true, execute body
            return exec_single(ifExpr->body);
        } else if (ifExpr->elseStmt) {
            // It's false, run the else part
            return exec_single(ifExpr->elseStmt);
        }
    } else if (CAST_TO(node.get(), LoopAST*)) {
        // Loops
        auto loop = std::reinterpret_pointer_cast<LoopAST>(node);
        // Get the var
        auto var = state.get_var(loop->var);
        if (!var) {
            // Define a new var if it doesn't exist
            state.vars[loop->var] = std::make_shared<Val>(0);
            var = state.vars[loop->var];
        }
        // Get iter
        auto iter = loop->iter;
        if (!CAST_TO(iter.get(), CallAST*)) {
            throw std::runtime_error("Only range() is currently supported for loops");
        }
        auto call = std::reinterpret_pointer_cast<CallAST>(iter);
        std::string name = CAST_TO(call->expr.get(), VarAST*)->name;
        if (name != "range") {
            throw std::runtime_error("Only range() is currently supported for loops");
        }
        if (call->args.size() != 2) {
            badArgsErr(call->args.size(), 3);
        }
        // Do the loop
        auto start = eval(call->args[0]);
        auto end = eval(call->args[1]);
        var->type = 'i';
        for (int i = start->toInt(); i <= end->toInt(); i++) {
            // Set var
            var->numVal = i;
            // Run
            auto ret = exec_single(loop->body);
            if (ret)
                return ret;
        }
    } else if (CAST_TO(node.get(), ImportAST*)) {
        // Get file
        auto import = std::reinterpret_pointer_cast<ImportAST>(node);
        std::string name = eval(import->expr)->toStr();
        // Import
        stream_t stream = {""};
        stream.str = read_input(name);
        stream.name = name;
        auto ast = genAST(stream);
        // Run
        bool oldIsRepl = isRepl;
        isRepl = false;
        if (!exec(ast)) {
            isRepl = oldIsRepl;
            throw std::runtime_error("Failed to import "+name);
        }
        isRepl = oldIsRepl;
    } else if (CAST_TO(node.get(), ReturnAST*)) {
        // Return causes errors outside of function
        if (!inFunc) {
            throw std::runtime_error("Return outside of function");
        }
        // Return from func
        auto ret = std::reinterpret_pointer_cast<ReturnAST>(node);
        auto expr = ret->expr;
        return eval(expr);
    }
    return NULL;
}

Val Runner::call_func(std::shared_ptr<CallAST> &call) {
    std::string name = CAST_TO(call->expr.get(), VarAST*)->name;
    // Inbuilt functions
    if (name == "print") {
        std::vector<std::string> strings = {};
        for (auto &arg : call->args) {
            strings.push_back(eval(arg)->toStr());
        }
        for (std::string &str : strings) {
            printf("%s ", str.c_str());
        }
        puts("");
        return Val();
    } else if (name == "range") {
        return Val(
            "Lists don't exist yet, this function can only be used in loops"
        );
    }
    // Casting
    else if (name == "string") {
        if (call->args.size() != 1)
            badArgsErr(call->args.size(), 1);
        return eval(call->args.at(0))->toStr();
    } else if (name == "bool") {
        if (call->args.size() != 1)
            badArgsErr(call->args.size(), 1);
        return !!*eval(call->args.at(0));
    } else if (name == "int") {
        if (call->args.size() != 1)
            badArgsErr(call->args.size(), 1);
        Val ret = *eval(call->args.at(0));
        if (ret.type == 's')
            // Convert string to int
            return std::stoi(ret.strVal);
        return ret.toInt();
    } else if (name == "float") {
        if (call->args.size() != 1)
            badArgsErr(call->args.size(), 1);
        Val ret = *eval(call->args.at(0));
        if (ret.type == 's')
            // Convert string to float
            return std::stof(ret.strVal);
        return ret.toFloat();
    }
    // It's a normal function
    std::shared_ptr<FuncAST> func = state.get_func(name);
    // Check if it exists
    if (!func) {
        noFuncErr(name);
        return Val();
    }
    // Check args
    if (call->args.size() != func->args.size()) {
        badArgsErr(call->args.size(), func->args.size());
    }
    // All good, prep for execution
    bool oldFunc = inFunc;
    inFunc = true;
    // Lower scope
    state_t curState = state;
    state = {};
    state.outer_state = &curState;
    // Set args
    int argNo = 0;
    for (std::string &arg : func->args) {
        state.vars[arg] = eval(call->args[argNo]);
        argNo++;
    }
    // Execute
    std::shared_ptr<Val> ret = exec_body(func->body, false);
    // Clean up
    state = curState;
    inFunc = oldFunc;
    return ret ? *ret : Val();
}

std::shared_ptr<Val> Runner::exec_body(
    std::shared_ptr<StmtAST> &bodyStmt, bool _lower_scope /*= true*/
) {
    // Lower scope
    state_t curState = {};
    if (_lower_scope) {
        curState = state;
        state = {};
        state.outer_state = &curState;
    }
    // Exec body
    auto body = CAST_TO(bodyStmt.get(), BodyAST*);
    std::shared_ptr<Val> ret = NULL;
    for (std::shared_ptr<StmtAST> &node : body->exprs) {
        ret = exec_single(node);
        // It's the return val
        if (ret) break;
    }
    // Raise scope back up
    if (_lower_scope) {
        state = curState;
    }
    // Return
    return ret;
}

bool Runner::exec(std::vector<std::shared_ptr<StmtAST>> &ast) {
    try {
        for (auto &node : ast) {
            // Exec
            std::shared_ptr<Val> val = exec_single(node);
            if (val != NULL) {
                // Print in repl mode
                puts(val->toStr().c_str());
            }
        }
        return true;
    } catch (std::runtime_error const &error) {
        printf("Runtime Error: %s\n", error.what());
        return false;
    }
}

File: src/interpreter.rs
use std::collections::HashMap;
use std::path::PathBuf;
use std::fmt;
use std::io::Write;
use std::io;

use crate::common::IMPOSSIBLE_STATE;
use crate::parser::ASTNode;
use ASTNode::*;
use crate::import_file;
use crate::lexer::TokenType;
use crate::value::Value;

// Functies are burlap functions in rust (like print)
type Functie = fn(&mut Interpreter, Vec<Value>) -> Value;

// Sub struct for extentions/functies
pub struct Functies {
    builtin: HashMap<String, Functie>
}
impl fmt::Debug for Functies {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
        write!(f, "{{...}}")
    }
}

// Interpreter state
#[derive(Debug)]
pub struct Interpreter {
    pub is_repl: bool,
    pub has_err: bool,
    pub in_func: bool,
    // Global vars and functions (which are always global)
    is_global: bool,
    globals: HashMap<String, Value>,
    functions: HashMap<String, ASTNode>,
    functies: Functies,
    // Variables in the current scope
    // Each scope stacks
    var_names: Vec<String>,
    var_vals: Vec<Value>,
    var_min: usize,
    // Extentions
    pub extentions: Vec<String>,
    // Used for importing
    pub import_path: PathBuf,
}

impl Interpreter {
    // Init
    pub fn new(
        is_repl: bool, import_path: PathBuf, extentions: Vec<String>
    ) -> Interpreter {
        // Builtin functions
        let functies = Functies{builtin: HashMap::from([
            ("print".to_string(), sk_print as Functie),
            ("input".to_string(), sk_input as Functie),
            // Casts
            ("int".to_string(), sk_int as Functie),
            ("float".to_string(), sk_float as Functie),
            ("string".to_string(), sk_string as Functie),
        ])};
        Interpreter {
            is_repl, has_err: false, in_func: false,
            is_global: true, globals: HashMap::new(),
            functions: HashMap::new(), functies,
            var_names: Vec::new(), var_vals: Vec::new(),
            var_min: 0, extentions, import_path
        }
    }
    // Getting vars
    fn get_local(&self, name: &String) -> Value {
        // Gets a local var
        let mut index = self.var_names.len();
        while index > self.var_min {
            index -= 1;
            if &self.var_names[index] == name {
                return self.var_vals[index].clone();
            }
        }
        // Failed to get var, return an error
        return Value::Error(format!("no variable called \"{}\"", name));
    }
    fn get_global(&self, name: &String) -> Value {
        // Gets a var in the global scope
        return match self.globals.get(name) {
            Some(val) => val.clone(),
            _ => Value::Error(format!("no variable called \"{}\"", name))
        };
    }
    pub fn get_var(&self, name: &String) -> Value {
        // Try to get it in the normal scope first
        return match self.get_local(name) {
            // Not found, try global scope
            Value::Error(_) => self.get_global(name),
            // It's in the local scope
            x => x,
        }
    }
    // Create a variable
    pub fn make_var(
        &mut self, name: &String, val: Value, global: bool, force: bool
    ) -> bool {
        if global {
            // Check
            if let Value::Error(_) = self.get_global(name) {
            } else if force {} else {
                return false;
            }
            // Add
            self.globals.insert(name.clone(), val);
        } else {
            // Check
            if let Value::Error(_) = self.get_local(name) {
            } else if force {} else {
                return false;
            }
            // Add
            self.var_names.push(name.clone());
            self.var_vals.push(val);
        }
        return true;
    }
    // Set a variable
    pub fn set_local(&mut self, name: &String, val: Value) -> bool {
        // Sets a local var
        let mut index = self.var_names.len();
        while index > self.var_min {
            index -= 1;
            if &self.var_names[index] == name {
                // Now that I have it, set it
                self.var_vals[index] = val;
                return true;
            }
        }
        // Failed to get var
        return false;
    }
    pub fn set_global(&mut self, name: &String, val: Value) -> bool {
        // Sets a global value
        if !self.globals.contains_key(name) {
            return false;
        }
        self.globals.insert(name.clone(), val);
        return true;
    }
    pub fn set_var(&mut self, name: &String, val: Value, global: bool) -> bool {
        // Sets a varible, returns true if it was successful
        return if global {
            self.set_global(name, val)
        } else {
            self.set_local(name, val)
        };
    }
    pub fn change_var(&mut self, name: &String, val: Value) -> bool {
        // Changes a varible to a diffrent value
        if !self.set_local(name, val.clone()) {
            return self.set_global(name, val)
        }
        return true;
    }
    // Raise/lower scope
    pub fn lower_scope(&mut self, call: bool) -> (bool, usize, usize) {
        // Lowers the scope
        // Impossible for a lowered scope to be global
        let old_global = self.is_global;
        self.is_global = false;
        // Functions can't access higher yet non-global scopes
        let old_min = self.var_min;
        if call {
            self.var_min = self.var_names.len();
        }
        // Where the vars need to be cut off at
        let old_top = self.var_names.len();
        // Return data needed to raise the dead/scope
        return (old_global, old_min, old_top);
    }
    pub fn raise_scope(&mut self, (old_global, old_min, old_top): (bool, usize, usize)) {
        // Raises the scope back
        self.is_global = old_global;
        self.var_min = old_min;
        // Remove new vars
        while old_top < self.var_names.len() {
            self.var_names.pop();
            self.var_vals.pop();
        }
    }
    fn bad_args(&self, name: &String, got: usize, need: usize) -> Value {
        if got > need {
            return Value::Error(
                format!("too mamy args for {} (got {} need {})", name, got, need)
            );
        }
        return Value::Error(
            format!("too few args for {} (got {} need {})", name, got, need)
        );
    }
    // Call a function
    pub fn call(&mut self, name: &String, args: &Vec<Value>) -> Value {
        // Builtin functions
        match (&self).functies.builtin.get(name) {
            Some(functie) => { return functie(self, args.clone()); },
            _ => ()
        }
        // Try to get the function from the name
        let function: &ASTNode;
        match (&self).functions.get(name) {
            Some(val) => function = val,
            _ => { return Value::Error(format!("no function called \"{}\"", name)); }
        };
        // Get the args and body
        let arg_names: Vec<String>;
        let body: ASTNode;
        if let ASTNode::FunctiStmt(_name, named_args, fbody) = function {
            arg_names = named_args.clone();
            body = (**fbody).clone();
        } else {
            // Impossible!
            return Value::Error(IMPOSSIBLE_STATE.to_string());
        }
        // Lower scope
        let old_func = self.in_func;
        self.in_func = true;
        let scope_data = self.lower_scope(true);
        // Push the args
        let mut index = 0;
        for arg_name in &arg_names {
            if index >= args.len() {
                // Error
                self.in_func = old_func;
                self.raise_scope(scope_data);
                return self.bad_args(name, args.len(), arg_names.len());
            }
            // Push arg
            self.make_var(&arg_name, args[index].clone(), false, true);
            index += 1;
        }
        // Check for too many args
        if index != args.len() {
            self.in_func = old_func;
            self.raise_scope(scope_data);
            return self.bad_args(name, args.len(), arg_names.len());
        }
        // Run body
        let ret = exec(self, &body);
        // Raise scope
        self.in_func = old_func;
        self.raise_scope(scope_data);
        return ret;
    }
}

// Builtin Functions (prefixed with 'sk_')
// Print
fn sk_print(interpreter: &mut Interpreter, args: Vec<Value>) -> Value {
    if interpreter.extentions.contains(&"va-print".to_string()) {
        // VA print extention
        for i in args {
            print!("{} ", i.to_string());
        }
        println!("");
    } else if args.len() != 1 {
        // Invalid args
        return interpreter.bad_args(&"print".to_string(), args.len(), 1);
    } else {
        // Normal printing
        println!("{}", args[0].to_string());
    }
    return Value::None;
}

// Input
fn sk_input(interpreter: &mut Interpreter, args: Vec<Value>) -> Value {
    if args.len() != 1 {
        // Invalid args
        return interpreter.bad_args(&"input".to_string(), args.len(), 1);
    }
    // Print the prompt
    print!("{}", args[0].to_string());
    let _ = std::io::stdout().flush();
    // Get input
    let mut buffer = String::new();
    return match io::stdin().read_line(&mut buffer) {
        Err(_) => Value::Str("".to_string()),
        _ => Value::Str(buffer.trim_end().to_string())
    };
}

// Casting
// Int
fn sk_int(interpreter: &mut Interpreter, args: Vec<Value>) -> Value {
    if args.len() != 1 {
        // Invalid args
        return interpreter.bad_args(&"int".to_string(), args.len(), 1);
    }
    return Value::Int(args[0].to_int());
}

// Float
fn sk_float(interpreter: &mut Interpreter, args: Vec<Value>) -> Value {
    if args.len() != 1 {
        // Invalid args
        return interpreter.bad_args(&"float".to_string(), args.len(), 1);
    }
    return Value::Float(args[0].to_float());
}

// String
fn sk_string(interpreter: &mut Interpreter, args: Vec<Value>) -> Value {
    if args.len() != 1 {
        // Invalid args
        return interpreter.bad_args(&"string".to_string(), args.len(), 1);
    }
    return Value::Str(args[0].to_string());
}

// Eval (for expressions)
fn eval(interpreter: &mut Interpreter, node: &ASTNode) -> Value {
    return match node {
        // Normal boring values
        StringExpr(val) => Value::Str(val.clone()),
        NumberExpr(val) => Value::Int(val.clone()),
        DecimalExpr(val) => Value::Float(val.clone()),
        BoolExpr(val) => Value::Bool(val.clone()),
        NoneExpr => Value::None,
        // Vars/function
        VarExpr(val) => interpreter.get_var(val),
        CallExpr(name, args) => {
            // Eval args
            let mut vals: Vec<Value> = vec![];
            for arg in args {
                let new_arg = eval(interpreter, arg);
                if let Value::Error(_) = new_arg {
                    return new_arg;
                }
                vals.push(new_arg);
            }
            // Call
            interpreter.call(name, &vals)
        }
        // Unary
        UnaryExpr(unary, val) => {
            // "&**" is a tad cursed, but this is rust after all.
            let value = eval(interpreter, &**val);
            if let Value::Error(_) = value {
                return value;
            }
            // Do the op
            let ret = match unary {
                TokenType::Not => Value::Bool(!value.is_truthy()),
                TokenType::Minus => Value::Int(0) - value.clone(),
                // These ones are done separately
                TokenType::PlusPlus | TokenType::MinusMinus => Value::Null,
                _ => Value::Error(format!("unknown unary op ({:?})", unary))
            };
            if let TokenType::PlusPlus | TokenType::MinusMinus = unary {} else {
                return ret;
            }
            // Do ++/--
            let name: String;
            if let ASTNode::VarExpr(n) = &**val {
                name = n.clone();
            } else {
                return Value::Error("++/-- can only be used on variable".to_string());
            }
            let val: Value;
            if let TokenType::PlusPlus = unary {
                // Increment
                val = value + Value::Int(1);
                interpreter.change_var(&name, val.clone());
            } else {
                // Decrement
                val = value - Value::Int(1);
                interpreter.change_var(&name, val.clone());
            }
            // Return
            return val;
        },
        // Binops
        BinopExpr(lhs, op, rhs) => {
            // Var name for variable ops
            let name = if let ASTNode::VarExpr(n) = &**lhs {
                n.clone()
            } else { "".to_string() };
            // Get values
            let left = eval(interpreter, &**lhs);
            let right = eval(interpreter, &**rhs);
            if let Value::Error(_) = left {
                return left;
            }
            if let Value::Error(_) = right {
                return right;
            }
            // Do the binop
            match op {
                // Math
                TokenType::Plus => left + right,
                TokenType::Minus => left - right,
                TokenType::Times => left * right,
                TokenType::Div => left / right,
                TokenType::Modulo => left % right,
                // Eq/neq
                TokenType::EqualsEquals => Value::Bool(left.eq(right)),
                TokenType::NotEquals => Value::Bool(!left.eq(right)),
                // Gt/lt
                TokenType::Lt => Value::Bool(left.to_float() < right.to_float()),
                TokenType::Gt => Value::Bool(left.to_float() > right.to_float()),
                TokenType::LtEquals => Value::Bool(left.to_float() <= right.to_float()),
                TokenType::GtEquals => Value::Bool(left.to_float() >= right.to_float()),
                // Logical
                TokenType::And =>
                    Value::Bool(left.is_truthy() && right.is_truthy()),
                TokenType::Or =>
                    Value::Bool(left.is_truthy() || right.is_truthy()),
                TokenType::Xor =>
                    Value::Bool(left.is_truthy() != right.is_truthy()),
                // Variable
                TokenType::Equals => {
                    interpreter.set_var(&name, right, interpreter.is_global);
                    Value::None
                },
                TokenType::PlusEquals => {
                    let val = interpreter.get_var(&name);
                    if let Value::Error(_) = val {
                        return val;
                    }
                    interpreter.change_var(&name, val + right);
                    Value::None
                },
                TokenType::MinusEquals => {
                    let val = interpreter.get_var(&name);
                    if let Value::Error(_) = val {
                        return val;
                    }
                    interpreter.change_var(&name, val - right);
                    Value::None
                },
                TokenType::TimesEquals => {
                    let val = interpreter.get_var(&name);
                    if let Value::Error(_) = val {
                        return val;
                    }
                    interpreter.change_var(&name, val * right);
                    Value::None
                },
                TokenType::DivEquals => {
                    let val = interpreter.get_var(&name);
                    if let Value::Error(_) = val {
                        return val;
                    }
                    interpreter.change_var(&name, val / right);
                    Value::None
                },
                // Unknown
                _ => Value::Error(format!("unknown binop op ({:?})", op))
            }
        },
        // Default to none
        _ => Value::Error(format!("eval error (failed to evaluate {:?})", node)),
    };
}

fn exec(interpreter: &mut Interpreter, node: &ASTNode) -> Value {
    return match node {
        BodyStmt(body) => {
            // Lower scope
            let scope_data = interpreter.lower_scope(false);
            // Run each statement
            let mut ret: Value;
            for i in body {
                ret = exec(interpreter, i);
                if let Value::Error(_) = ret {
                    return ret;
                }
                if ret != Value::Null {
                    // Raise scope
                    interpreter.raise_scope(scope_data);
                    return ret;
                }
            }
            // Raise scope
            interpreter.raise_scope(scope_data);
            Value::Null
        },
        ReturnStmt(ret) => {
            // Return
            eval(interpreter, &**ret)
        },
        IfStmt(cond, body, elsestmt) => {
            // Execute either body or else body
            let condition = eval(interpreter, &**cond);
            if let Value::Error(_) = condition {
                return condition;
            }
            if condition.is_truthy() {
                exec(interpreter, &**body)
            } else {
                // If there isn't a else cond, this will run an empty body
                exec(interpreter, &**elsestmt)
            }
        },
        // Declaring vars
        LetStmt(name, val) => {
            let val = eval(interpreter, &**val);
            if let Value::Error(_) = val {
                return val;
            }
            if interpreter.make_var(name, val, interpreter.is_global, false) {
                Value::Null
            } else {
                Value::Error(format!("cannot redefine variable \"{}\"", name))
            }
        },
        // Declaring functions
        FunctiStmt(name, _args, _body) => {
            interpreter.functions.insert(name.clone(), node.clone());
            Value::Null
        },
        // Loops
        LoopStmt(var, iter, body) => {
            // Make the var if it doesn't exist
            if let Value::Error(_) = interpreter.get_var(var) {
                interpreter.make_var(var, Value::Int(0), interpreter.is_global, false);
            }
            // Check iter
            let (mut min, max): (i32, i32);
            if let CallExpr(name, args) = &**iter {
                // Check name
                if name != "range" {
                    return Value::Error(
                        "only range is currently supported for iterative loops"
                    .to_string());
                }
                // Check args
                if args.len() != 2 {
                    return interpreter.bad_args(&name, args.len(), 2);
                }
                // Get args
                min = eval(interpreter, &args[0]).to_int();
                max = eval(interpreter, &args[1]).to_int();
            } else {
                return Value::Error(
                    "only range is currently supported for iterative loops".to_string()
                );
            }
            // Do the loop
            while min <= max {
                interpreter.set_var(var, Value::Int(min), interpreter.is_global);
                exec(interpreter, &*body.clone());
                min += 1;
            }
            Value::Null
        },
        // While loop
        WhileStmt(cond, body) => {
            // It's really easy, just check cond and loop
            while eval(interpreter, &**cond).is_truthy() {
                exec(interpreter, &*body.clone());
            }
            Value::Null
        },
        // Import
        ImportStmt(name) => {
            // Get file name
            let fname = eval(interpreter, &**name);
            let name = fname.to_string();
            // Get the file path
            let mut path = interpreter.import_path.clone();
            path.push(name.clone());
            // Import
            if import_file(interpreter, &mut path) {
                Value::Null
            } else {
                Value::Error(format!("failed to import {}", name))
            }
        },
        _ => {
            // Expression, not statement
            let val = eval(interpreter, node);
            if let Value::Error(_) = val {
                return val;
            }
            if interpreter.is_repl && val != Value::Null && val != Value::None {
                println!("{}", val.to_string());
            }
            Value::Null
        }
    }
}

pub fn run(interpreter: &mut Interpreter, ast: Vec<ASTNode>) -> bool {
    for node in ast {
        let val = exec(interpreter, &node);
        if let Value::Error(msg) = val {
            println!("RuntimeError: {}", msg);
            return false;
        }
    }
    return true;
}
